admin/controllers/pupil.js

```diff
@@ -95,27 +81,142 @@ const postAddPupil = async (req, res, next) => {
   } catch (error) {
     next(error)
   }
-  res.redirect(`/school/pupil-register/lastName/true?hl=${pupil._id}`)
+  const pupilId = JSON.stringify([pupil._id])
+  res.redirect(`/school/pupil-register/lastName/true?hl=${pupilId}`)
 }
 
-const getAddMultiplePupils = async (req, res, next) => {
+const getAddMultiplePupils = (req, res, next) => {
   res.locals.pageTitle = 'Add multiple pupils'
+  const { hasError } = res
   try {
     req.breadcrumbs('Pupil Register', '/school/pupil-register/lastName/true')
     req.breadcrumbs(res.locals.pageTitle)
     res.render('school/add-multiple-pupils', {
-      breadcrumbs: req.breadcrumbs()
+      breadcrumbs: req.breadcrumbs(),
+      hasError
     })
   } catch (error) {
     next(error)
   }
 }
 
+const postAddMultiplePupils = async (req, res, next) => {
+  let school
+  try {
+    school = await School.findOne({_id: req.user.School}).exec()
+    if (!school) {
+      throw new Error(`School [${req.body.school}] not found`)
+    }
+  } catch (error) {
+    return next(error)
+  }
+  const uploadFile = req.files.csvTemplateFile
+  let csvData = []
+  const stream = fs.createReadStream(uploadFile.file)
+  const csvStream = csv()
+    .on('data', (data) => { csvData.push(data) })
+    .on('end', async() => {
+      // Remove error column and headers from data
+      if (csvData.some(p => p[6])) csvData.map((r) => r.splice(6, 1))
+      let headers = csvData.shift(0)
+      // validate each pupil
+      const pupils = []
+      csvData = await Promise.all(csvData.map(async(p) => {
+        const pupil = new Pupil({
+          school: school._id,
+          upn: p[3],
+          foreName: p[0],
+          lastName: p[2],
+          middleNames: p[1],
+          gender: p[5],
+          dob: p[4],
+          pin: null,
+          pinExpired: false
+        })
+        try {
+          const dob = p[4].split('/')
+          const pupilData = Object.assign({
+            'dob-day': dob[1],
+            'dob-month': dob[0],
+            'dob-year': dob[2]
+          }, pupil._doc)
+          await validatePupil(pupil, pupilData)
+        } catch (err) {
+          p[6] = []
+          Object.keys(err.errors).forEach((e) => p[6].push(err.errors[e].replace(/ \\"/g, '')))
+          p[6] = p[6].join(', ')
+        }
+        pupils.push(pupil)
+        return p
+      }))
+      // Generate csv with errors
+      if (csvData.some(p => p[6])) {
+        const errorsCsv = []
+        headers.push('Errors')
+        errorsCsv.push(headers)
+        csvData.forEach((p) => errorsCsv.push(p))
+        const writeToString = promisify(csv.writeToString)
+        const cvsStr = await writeToString(errorsCsv, { headers: true })
+        // Upload csv to Azure
+        try {
+          const remoteFilename = moment().format('YYYYMMDDHHmmss') + '-' + 'error.csv'
```

Comment:
@GeorgeChatzigiannis  i would use something more unique here - it is entirely possible that 2 or more users could be uploading files that contain errors at the same time.

Reply:
@GuyHarwood @GeorgeChatzigiannis - as node often aligns to unix I had assumed there would be a `mktemp` available.  There isn't, but there is a `mkdtemp` which will create a temp directory.  Otherwise there is a package for it: https://www.npmjs.com/package/mktemp

Reply:
@GuyHarwood I added school id in the beginning of the file name. 
Let me know if we need further uniqueness